### PR TITLE
As-3-updates

### PR DESCRIPTION
**Pull request by KayedObeidat**
1- I moved the API's/DB calls to a single file and call them inside the routes.
2- I add the routes validations
3- I used the app.js file because user.test.js doesn't work without this file.

### CODE REVIEW

File: src/routes/user.js [hunk 1]
```diff
@@ -2,25 +2,19 @@ const express = require('express')
...
-    }catch(e){
-        res.status(400).send(e)
-    }
-})
+
+router.get('/users',
+    services.readUsers
+)
 
 router.get('/users/:id', async(req, res) => {
```
Comment (KayedObeidat) [inline src/routes/user.js:19]: I tried to validate the ID but, unfortunately, it doesn't workm

File: src/routes/user.js [hunk 2]
```diff
@@ -34,32 +28,56 @@ router.get('/users/:id', async(req, res) => {
     }
 })
 
-router.patch('/users/update/:id', async(req, res) => {
-     const updates = Object.keys(req.body)
-     const allwoedUpdates = ['name', 'password', 'age']
-     const isValidOperation = updates.every((update) => {
-         return allwoedUpdates.includes(update)
-     })
+// router.patch('/users/update/:id', async(req, res) => {
```
Comment (KayedObeidat) [inline src/routes/user.js]: Whenever I try to validate the ID, the terminal gave me something like req.params is undefiend

File: src/services/user.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+const mongoose = require('mongoose')
+const validator = require('validator')
+const { body, validationResult } = require('express-validator')
+
+// exports.validate = [
+//     body('email').isEmail(),
+  
+//     body('password').isLength({ min: 7 })
+// ]
+exports.validate = (method) => {
```
Comment (KayedObeidat) [inline src/services/user.js]: To validate name, email, and password

File: src/services/user.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+//   exports.readUser = (req, res) => {
+//       const errors = validationResult(req)
+//       if(!errors.isEmpty()){
+//           return res.status(404).send({ errors: errors.array() })
+//       }
+//       User.findById(req.params.id).then(users => res.status(200).send(user) )
+//   }
+
+exports.createUser = (req, res) => {
```
Comment (KayedObeidat) [inline src/services/user.js:6]: Creating a new user

File: src/services/user.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        User.create({
+                name: req.body.name,
+                email: req.body.email,
+                password: req.body.password,
+                age: req.body.age
+        }).then(user => res.status(201).json(user))
+    }
+
+
+exports.readUsers = (req, res) => {
```
Comment (KayedObeidat) [inline src/services/user.js:21]: Fetch users from the database

File: src/services/user.js (hunk 3 above)
Comment (Boltmerz) [reply src/services/user.js]: @KayedObeidat validation functions should be in the middleware or validation folder NOT in the services. Services are for DB & external APIs calls only.

File: src/routes/user.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        res.status(400).send(e)
+    }
+})
+
+router.patch('/users/update/:id', async(req, res) => {
+    const updates = Object.keys(req.body)
+    const allwoedUpdates = ['name', 'password', 'age']
+    const isValidOperation = updates.every((update) => {
+        return allwoedUpdates.includes(update)
+    })
```
Comment (Boltmerz) [inline src/routes/user.js:36]: Move 32-36 to a validation middleware function.

File: src/app.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+const User = require('./models/user')
+const userRouter = require('./routes/user')
+
+
+const app = express()
+app.use(express.json())
+app.use(userRouter)
+
+
+module.exports = app
```
Comment (Boltmerz) [inline src/app.js:17]: Merge changes here with index.js and keep index.js. 
If your test fails without app.js , you need to update the test and we can look at that later.

File: src/index.js [hunk 8]
```diff
@@ -1,16 +1,15 @@
 const express = require('express')
 require('./dataBase/mongoose')
-
-const app = express()
+const User = require('./models/user')
```
Comment (Boltmerz) [inline src/index.js:3]: this is not used!

File: src/routes/user.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    if(!isValidOperation){
+        return res.status(400).send({ Error: 'Invalid Updates' })
+    }
+    try{
+        const user = await User.findByIdAndUpdate(req.params.id)
+        updates.forEach((update) => {
+        user[update] = req.body[update]
```
Comment (Boltmerz) [inline src/routes/user.js:45]: Please read the documentation for findByIdAndUpdate() and re-write the code

File: test/user.test.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+})
+
+test('Should update user', async() => {
+        await request(app)
+            .patch(`/users/update/${userOneId}`)
+            .send({ name: 'kayed' })
+            .expect(200)
+            const user = await User.findByIdAndUpdate(userOneId)
+            expect(user.name).equal('kayed')
+})
```
Comment (Boltmerz) [inline test/user.test.js:41]: This function will update DB. Please see my previous comments.

**Review by Boltmerz — CHANGES_REQUESTED** [review]
Please read my previous comments.
Also add unit tests for controllers and services and DB should not be called when unit tests execute.
Read about jest.mock()

**Pull request merged**